snuba/state/rate_limit.py

```diff
@@ -111,26 +116,60 @@ def rate_limit(
     now = time.time()
     bypass_rate_limit, rate_history_s = state.get_configs(
         [("bypass_rate_limit", 0), ("rate_history_sec", 3600)]
+        #                               ^ number of seconds we keep the timestamps
     )
     assert isinstance(rate_history_s, (int, float))
 
+    # we don't really do this but you *can* skip rate limit
     if bypass_rate_limit == 1:
         yield None
         return
 
     pipe = rds.pipeline(transaction=False)
-    pipe.zremrangebyscore(
-        bucket, "-inf", "({:f}".format(now - rate_history_s)
-    )  # cleanup
+    # cleanup old queries: remove everything between [-inf and now - rate_history_s]
+    pipe.zremrangebyscore(bucket, "-inf", "({:f}".format(now - rate_history_s))
+
+    # Now for the tricky bit:
+    # ======================
+    # we add the query's *deadline* to the sorted set of timestamps, therefore
+    # labeling its execution as in the future.
+
+    # We consider all queries with timestamps in the future to be executing *right now*
+    # Example:
+
+    # now = 100
+    # max_query_duration_s = 30
+    # rate_lookback_s = 10
+    # sorted_set (timestamps only for clarity) = [91, 94, 97, 103, 105, 130]
+
+    # EXPLANATION:
+    # ===========
+
+    # queries that have finished running
+    # (in this example there are 3 queries in the last 10 seconds
+    #  thus the per second rate is 3/10 = 0.3)
+    #      |
+    #      v
+    #  -----------              v--- our current query, vaulted into the future
+    #  [91, 94, 97, 103, 105, 130]
+    #               -------------- < - queries currently running
+    #                                (this tells us how many queries are
+    #                                   running concurrently; in this case 3)
+    # .            ^
+    #              | current time
+
     pipe.zadd(bucket, now + state.max_query_duration_s, query_id)  # type: ignore
     if rate_limit_params.per_second_limit is None:
         pipe.exists("nosuchkey")  # no-op if we don't need per-second
     else:
-        pipe.zcount(bucket, now - state.rate_lookback_s, now)  # get historical
+        # count queries that have finished for the per-second rate
+        pipe.zcount(bucket, now - state.rate_lookback_s, now)
     if rate_limit_params.concurrent_limit is None:
         pipe.exists("nosuchkey")  # no-op if we don't need concurrent
     else:
-        pipe.zcount(bucket, "({:f}".format(now), "+inf")  # get concurrent
+        # count the amount queries in the "future" which tells us the amount
+        # of concurrent queries
+        pipe.zcount(bucket, "({:f}".format(now), "+inf")
```

Comment:
The code is pretty self explanatory here.

Reply:
I disagree. This comment ties the diagram at the top to the actual code operation. I think this should stay.